### PR TITLE
feat: Territory defense level sorting

### PR DESCRIPTION
**Pull request by ryanzhoudev**
Textures aren't really the most centered, but I will fix all the textures in a separate PR

### CODE REVIEW

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 1]
```diff
@@ -321,4 +355,32 @@ private void renderPois(PoseStack poseStack, int mouseX, int mouseY) {
...
+    public enum TerritoryDefenseLevel {
+        Off(ChatFormatting.GRAY + "Off"),
+        VeryLow(ChatFormatting.DARK_GREEN + "Very Low"),
+        Low(ChatFormatting.GREEN + "Low"),
+        Medium(ChatFormatting.YELLOW + "Medium"),
+        High(ChatFormatting.RED + "High"),
+        VeryHigh(ChatFormatting.DARK_RED + "Very High");
+
+        private final String asColoredString;
+        private static final TerritoryDefenseLevel[] VALUES = values();
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: Is this needed?

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 2]
```diff
@@ -321,4 +355,32 @@ private void renderPois(PoseStack poseStack, int mouseX, int mouseY) {
     public boolean isResourceMode() {
         return resourceMode;
     }
+
+    public TerritoryDefenseLevel getTerritoryDefenseFilterLevel() {
+        return territoryDefenseFilterLevel;
+    }
+
+    public enum TerritoryDefenseLevel {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: Move this into its own file

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 3]
```diff
@@ -321,4 +355,32 @@ private void renderPois(PoseStack poseStack, int mouseX, int mouseY) {
...
+        Off(ChatFormatting.GRAY + "Off"),
+        VeryLow(ChatFormatting.DARK_GREEN + "Very Low"),
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: I would rather have all of these capitalized. `VERY_LOW`

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 4]
```diff
@@ -27,12 +27,30 @@
...
 import net.minecraft.network.chat.TranslatableComponent;
 import org.lwjgl.glfw.GLFW;
 
 public class GuildMapScreen extends AbstractMapScreen {
     private boolean resourceMode = false;
+    private TerritoryDefenseLevel territoryDefenseFilterLevel = TerritoryDefenseLevel.Off;
+
+    private BasicTexturedButton territoryDefenseFilterButton;
+
+    private List<Component> getTerritoryDefenseFilterButtonTooltip(TerritoryDefenseLevel tdfl) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: Maybe the enum could contain this code?

File: common/src/main/java/com/wynntils/gui/widgets/BasicTexturedButton.java [hunk 5]
```diff
@@ -60,4 +61,12 @@ public void onPress() {}
 
     @Override
     public void updateNarration(NarrationElementOutput narrationElementOutput) {}
+
+    public void editTooltip(List<Component> newTooltip) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/widgets/BasicTexturedButton.java]: line 31 (in ctor) could use this method too, so it is generalized

File: common/src/main/java/com/wynntils/gui/widgets/BasicTexturedButton.java [hunk 6]
```diff
@@ -60,4 +61,12 @@ public void onPress() {}
...
+        tooltip = ComponentUtils.wrapTooltips(newTooltip, 250);
+    }
+
+    public List<Component> getTooltip() {
+        return Collections.unmodifiableList(tooltip);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/widgets/BasicTexturedButton.java]: What if we made the tooltip unmodifiable in `editTooltip`?

File: common/src/main/java/com/wynntils/wynn/model/map/poi/TerritoryPoi.java [hunk 7]
```diff
@@ -43,6 +43,19 @@ public TerritoryPoi(TerritoryProfile territoryProfile, TerritoryInfo territoryIn
     @Override
     public void renderAt(
             PoseStack poseStack, float renderX, float renderZ, boolean hovered, float scale, float mapZoom) {
+        if (McUtils.mc().screen instanceof GuildMapScreen guildMapScreen) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/wynn/model/map/poi/TerritoryPoi.java]: Shouldn't we filter out what pois to render in `GuildMapScreen`?

File: common/src/main/resources/assets/wynntils/lang/en_us.json [hunk 8]
```diff
@@ -770,6 +770,16 @@
...
+  "screens.wynntils.guildMap.cycleDefenseFilter.description1": "Click here to cycle the territory defense filter",
+  "screens.wynntils.guildMap.cycleDefenseFilter.description2": "Hold SHIFT to filter higher, CTRL to filter lower",
+  "screens.wynntils.guildMap.cycleDefenseFilter.description3": "Current value: %s",
+  "screens.wynntils.guildMap.cycleDefenseFilter.high": "§cHigh",
+  "screens.wynntils.guildMap.cycleDefenseFilter.low": "§aLow",
+  "screens.wynntils.guildMap.cycleDefenseFilter.medium": "§eMedium",
+  "screens.wynntils.guildMap.cycleDefenseFilter.name": "Cycle Territory Defense Filter",
+  "screens.wynntils.guildMap.cycleDefenseFilter.off": "§7Off",
+  "screens.wynntils.guildMap.cycleDefenseFilter.veryHigh": "§cVery High",
+  "screens.wynntils.guildMap.cycleDefenseFilter.veryLow": "§aVery Low",
```
Comment (kristofbolyai) [inline common/src/main/resources/assets/wynntils/lang/en_us.json]: These are unused (except `"screens.wynntils.guildMap.cycleDefenseFilter.name"`)? Are these meant to be used in the enum? If so, dont include the color codes in the localization

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 9]
```diff
@@ -73,6 +91,22 @@ protected void init() {
...
+                        - 6),
+                16,
+                16,
+                Texture.MAP_ADD_BUTTON, // TODO: Add new cycle texture
+                (b) -> {
+                    territoryDefenseFilterLevel = territoryDefenseFilterLevel.next();
+                    territoryDefenseFilterButton.editTooltip(
+                            getTerritoryDefenseFilterButtonTooltip(territoryDefenseFilterLevel));
+                },
+                getTerritoryDefenseFilterButtonTooltip(territoryDefenseFilterLevel)));
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: At some point we should create functionality to automatically add buttons at the right location

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: `width / 2 - Texture.MAP_BUTTONS_BACKGROUND.width() / 2 + 6 + 20 * buttonSlot`, already made it so you don't have to guess.

File: common/src/main/java/com/wynntils/gui/widgets/BasicTexturedButton.java (hunk 6 above)
Comment (ryanzhoudev) [reply common/src/main/java/com/wynntils/gui/widgets/BasicTexturedButton.java]: This was actually just for debug and didn't need to exist

File: common/src/main/resources/assets/wynntils/lang/en_us.json (hunk 8 above)
Comment (ryanzhoudev) [reply common/src/main/resources/assets/wynntils/lang/en_us.json]: Not meant to be used, was originally going to use them but the GuildResourceValues enum doesn't use them so the current comparison will break if someone switches languages

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 10]
```diff
@@ -122,6 +142,22 @@ protected void renderPois(
 
         List<Poi> filteredPois = getRenderedPois(pois, textureBoundingBox, poiScale, mouseX, mouseY);
 
+        if (territoryDefenseFilterLevel != TerritoryDefenseLevel.OFF) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: You should do this in `renderPois` instead.

File: common/src/main/resources/assets/wynntils/lang/en_us.json [hunk 11]
```diff
@@ -772,6 +772,10 @@
   "screens.wynntils.characterSelection.playButton.description": "You can also double click to play",
   "screens.wynntils.characterSelection.playButton.play": "[>] Play with the selected character",
   "screens.wynntils.gearViewer.viewStats": "View Player's Statistics",
+  "screens.wynntils.guildMap.cycleDefenseFilter.description1": "Click here to cycle the territory defense filter",
```
Comment (kristofbolyai) [inline common/src/main/resources/assets/wynntils/lang/en_us.json]:
```suggestion
  "screens.wynntils.guildMap.cycleDefenseFilter.description1": "Click here to cycle the territory defense filter.",
```

File: common/src/main/resources/assets/wynntils/lang/en_us.json [hunk 12]
```diff
@@ -772,6 +772,10 @@
...
+  "screens.wynntils.guildMap.cycleDefenseFilter.description2": "Hold SHIFT to filter higher, CTRL to filter lower",
```
Comment (kristofbolyai) [inline common/src/main/resources/assets/wynntils/lang/en_us.json]:
```suggestion
  "screens.wynntils.guildMap.cycleDefenseFilter.description2": "Hold SHIFT to filter higher, CTRL to filter lower.",
```

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java (hunk 10 above)
Comment (ryanzhoudev) [reply common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: This is in renderPois()

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: The other one, line 300 ish

File: common/src/main/java/com/wynntils/wynn/model/map/TerritoryDefenseLevel.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.wynn.model.map;
+
+import net.minecraft.ChatFormatting;
+
+public enum TerritoryDefenseLevel {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/wynn/model/map/TerritoryDefenseLevel.java]: I just noticed. `GuildResourceValues` is basically the same enum as this one. Can you remove this?

File: common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java [hunk 14]
```diff
@@ -309,6 +352,41 @@ private void renderHoveredTerritoryInfo(PoseStack poseStack) {
     private void renderPois(PoseStack poseStack, int mouseX, int mouseY) {
         List<Poi> pois = Managers.Territory.getTerritoryPoisFromAdvancement();
```
Comment (magicus) [inline common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: You should send in territoryDefenseFilterType to getTerritoryPoisFromAdvancement, and let the filtering happen there. Or maybe create a new `getFilteredTerritoryPois` in the  TerritoryManager.

Comment (magicus) [reply common/src/main/java/com/wynntils/gui/screens/maps/GuildMapScreen.java]: Or maybe keep some of the logic here, make a `getLevelFilteredTerritoryPois` which takes a `Predicate<Integer>` as a filter, and you can call it like 

```
getLevelFilteredTerritoryPois(level -> switch (territoryDefenseFilterType) {
                            case LOWER -> level <= territoryDefenseFilterLevel.getLevel();
                            // etc
```

(and in the implementation, use `!filter.apply(...)` so you don't have to invert the conditions here.

**Review by magicus — APPROVED** [review]

**Review by magicus — APPROVED** [review]

**Pull request merged**